### PR TITLE
fix(Scripts/ZulAman): Nalorakk can't able to trigger instance events when death

### PR DESCRIPTION
**Pull request by kissingers**
When this boss died, a notification should be given to the instance even, so the instance even can decide whether or not to summon a NPC,  also can  counting the number of boss kills to decide whether to open some doors or not.
Lack of this notification will result in players not being able to get the items in the box given by the NPC. and can't open the door after 4 bosses(if all doors are fixed).

## Changes Proposed:

This PR proposes changes to:
-  [ ] Core (units, players, creatures, game systems).
-  [x] Scripts (bosses, spell scripts, creature scripts).
-  [ ] Database (SAI, creatures, etc).

## Issues Addressed:

will fix some of the issues, but not all.
https://github.com/azerothcore/azerothcore-wotlk/issues/19048
- Closes 

## SOURCE:

The changes have been validated through:
- [ ] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [ ] Sniffs (remember to share them with the open source community!)
- [ ] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

## Tests Performed:

This PR has been:
- [ ] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [ ] This pull request requires further testing and may have edge cases to be tested.

## How to Test the Changes:

- [ ] This pull request can be tested by following the reproduction steps provided in the linked issue
- [ ] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.

1.
2.
3.

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/scripts/EasternKingdoms/ZulAman/boss_nalorakk.cpp [hunk 1]
```diff
@@ -346,6 +346,15 @@ struct boss_nalorakk : public BossAI
         }
         return true;
     }
+
+    void JustDied(Unit* killer) override
+    {
+        BossAI::JustDied(killer);
+        Talk(SAY_DEATH);
+        if (instance)
+            instance->SetData(DATA_NALORAKKEVENT, DONE);
```
Comment (PkllonG) [inline src/server/scripts/EasternKingdoms/ZulAman/boss_nalorakk.cpp:355]: 删掉
delete
```
        if (instance)
            instance->SetData(DATA_NALORAKKEVENT, DONE);
```

**Comment by kissingers** [comment]
if not do SetData(DATA_NALORAKKEVENT, DONE),   the  instance_zulaman.cpp  here will never run, without SummonHostage(0) will lost the first box.

                case DATA_NALORAKKEVENT:
                    m_auiEncounter[DATA_NALORAKKEVENT] = data;
                    if (data == DONE)
                    {
                        if (QuestMinute)
                        {
                            QuestMinute += 15;
                            DoUpdateWorldState(WORLDSTATE_TIME_TO_SACRIFICE, QuestMinute);
                        }
                        SummonHostage(0);
                        SaveToDB();
                    }
                    break;

**Comment by PkllonG** [comment]
> if not do SetData(DATA_NALORAKKEVENT, DONE), the instance_zulaman.cpp here will never run, without SummonHostage(0) will lost the first box.
> 
> ```
>             case DATA_NALORAKKEVENT:
>                 m_auiEncounter[DATA_NALORAKKEVENT] = data;
>                 if (data == DONE)
>                 {
>                     if (QuestMinute)
>                     {
>                         QuestMinute += 15;
>                         DoUpdateWorldState(WORLDSTATE_TIME_TO_SACRIFICE, QuestMinute);
>                     }
>                     SummonHostage(0);
>                     SaveToDB();
>                 }
>                 break;
> ```

OK My bad

**Comment by kissingers** [comment]
next try fix instance problem. the most problem is not right ++ bosscount,  we just ++ boss for real boss but not all done event, so   I change to below, now all ok and no any problem

        void SetData(uint32 type, uint32 data) override
        {
            switch (type)
            {
            case DATA_GONGEVENT:
                m_auiEncounter[DATA_GONGEVENT] = data;
                if (data == DONE)
                    QuestMinute = 21;
                break;
            case DATA_NALORAKKEVENT:
                m_auiEncounter[DATA_NALORAKKEVENT] = data;
                if (data == DONE)
                {
                    ++BossKilled;
                    if (QuestMinute)
                    {
                        QuestMinute += 15;
                        DoUpdateWorldState(WORLDSTATE_TIME_TO_SACRIFICE, QuestMinute);
                        SummonHostage(0);
                    }
                }
                break;
            case DATA_AKILZONEVENT:
                m_auiEncounter[DATA_AKILZONEVENT] = data;
                HandleGameObject(AkilzonDoorGUID, data != IN_PROGRESS);
                if (data == DONE)
                {
                    ++BossKilled;
                    if (QuestMinute)
                    {
                        QuestMinute += 10;
                        DoUpdateWorldState(WORLDSTATE_TIME_TO_SACRIFICE, QuestMinute);
                        SummonHostage(1);
                    }
                }
                break;
            case DATA_JANALAIEVENT:
                m_auiEncounter[DATA_JANALAIEVENT] = data;
                if (data == DONE)
                {
                    ++BossKilled;
                    if (QuestMinute)
                        SummonHostage(2);
                }

                break;
            case DATA_HALAZZIEVENT:
                m_auiEncounter[DATA_HALAZZIEVENT] = data;
                HandleGameObject(HalazziDoorGUID, data != IN_PROGRESS);
                if (data == DONE)
                {
                    ++BossKilled;
                    if (QuestMinute)
                        SummonHostage(3);
                }
                break;
            case DATA_HEXLORDEVENT:
                m_auiEncounter[DATA_HEXLORDEVENT] = data;
                if (data == IN_PROGRESS)
                    HandleGameObject(HexLordGateGUID, false);
                else if (data == NOT_STARTED)
                    CheckInstanceStatus();
                else if (data == DONE)
                    ++BossKilled;
                break;
            case DATA_ZULJINEVENT:
                m_auiEncounter[DATA_ZULJINEVENT] = data;
                HandleGameObject(ZulJinDoorGUID, data != IN_PROGRESS);
                SaveToDB();
                break;
            case DATA_CHESTLOOTED:
                ++ChestLooted;
                SaveToDB();
                break;
            case TYPE_RAND_VENDOR_1:
                RandVendor[0] = data;
                break;
            case TYPE_RAND_VENDOR_2:
                RandVendor[1] = data;
                break;
            }

            if (data == DONE)
            {
                if (QuestMinute && BossKilled >= DATA_HALAZZIEVENT)
                {
                    QuestMinute = 0;
                    DoUpdateWorldState(WORLDSTATE_SHOW_TIMER, 0);
                }
                CheckInstanceStatus();
                SaveToDB();
            }
        }

if use 80 level way, maybe delete SummonHostage(3) .

**Comment by kissingers** [comment]
and if fix the door sql , the chage also can right ctrl the door, olny need change door  186304 to 186306

GO_GATE_ZULJIN = 186306,

and load  https://github.com/azerothcore/azerothcore-wotlk/issues/19048 sql

**Comment by kissingers** [comment]
Tested
Increase the countdown time correctly and can get package rewards. if after the countdown time will get no rewards, which is consistent with Blizzard's behavior

**Review by kissingers — COMMENTED** [review]
Reviewed

**Comment by kissingers** [comment]
@elthehablo 
After the PR https://github.com/azerothcore/azerothcore-wotlk/pull/19893 ,the boss can not loot the box and can not add 15m time
So if this pr not a good way to fix the problem, might approve this pr becasue now can not normal operate with the instance. 
Or fix the problem some other way as soon as possible, as you referenced "more in the coming months so we can work on improving this further"

**Review by elthehablo — APPROVED** [review]

File: src/server/scripts/EasternKingdoms/ZulAman/boss_nalorakk.cpp [hunk 2]
```diff
@@ -346,6 +346,15 @@ struct boss_nalorakk : public BossAI
         }
         return true;
     }
+
+    void JustDied(Unit* killer) override
+    {
+        BossAI::JustDied(killer);
+        Talk(SAY_DEATH);
+        if (instance)
```
Comment (Nefertumm) [inline src/server/scripts/EasternKingdoms/ZulAman/boss_nalorakk.cpp:354]: This check is unnecessary, instance would never be null.

**Pull request merged**